java-client/session/src/main/java/io/deephaven/client/impl/UnaryGrpcFuture.java

```diff
@@ -0,0 +1,93 @@
+package io.deephaven.client.impl;
+
+import io.grpc.stub.ClientCallStreamObserver;
+import io.grpc.stub.ClientResponseObserver;
+import io.grpc.stub.StreamObserver;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+
+/**
+ * ClientResponseObserver implementation that can map results into a CompletableFuture. To keep generics under control,
+ * has two factory methods, {@link #of(Object, BiConsumer, Function)} and {@link #ignoreResponse(Object, BiConsumer)}.
+ * 
+ * @param <ReqT> The type of the gRPC request
+ * @param <RespT> The type of the gRPC response
+ * @param <V> The type of the result to make available in the future
+ */
+public class UnaryGrpcFuture<ReqT, RespT, V> implements ClientResponseObserver<ReqT, RespT> {
+    /**
+     * Factory method to take a request instance, a method reference to a unary call, and a mapping function to
+     * transform the result before supplying it to the future.
+     *
+     * @param request the gRPC request value
+     * @param call the unary gRPC call reference/lambda
+     * @param mapping a mapping function to transform the response into a result
+     * @param <ReqT> The type of the gRPC request
+     * @param <RespT> The type of the gRPC response
+     * @param <V> The type of the result to make available in the future
+     * @return A CompletableFuture that will succeed, fail, or be canceled. If it succeeds, will contain a result from
+     *         the mapping function.
+     */
+    public static <ReqT, RespT, V> CompletableFuture<V> of(ReqT request, BiConsumer<ReqT, StreamObserver<RespT>> call,
+            Function<RespT, V> mapping) {
+        UnaryGrpcFuture<ReqT, RespT, V> observer = new UnaryGrpcFuture<>(mapping);
+        call.accept(request, observer);
+        return observer.future();
+    }
```

Comment:
What are your thoughts on having the callers always get `CompletableFuture<RespT>` back here (ie, no `mapping` arg), and they can call `.thenApply(...)` as appropriate?

Reply:
I actually started off that way, as it seems nicer to be "fluent" like that.  The problem there is that if they return the result of `.thenApply()` or whatnot, then the caller is left with the downstream CF, and canceling that won't cancel the upstream one (so won't stop the grpc call).

Reply:
Good call. While the caller _could_ still add their own `.whenComplete(...)` and pass along the cancel, that's a bunch of extra machinery I agree we don't really want.